src/tpm2-tss-engine-ciphers.c

```diff
@@ -0,0 +1,527 @@
+ * Copyright 2019, Schneider-Electric
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ * this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * 3. Neither the name of tpm2-tss-engine nor the names of its contributors
+ * may be used to endorse or promote products derived from this software
+ * without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ ******************************************************************************/
+
+#include <string.h>
+
+#include <openssl/engine.h>
+#include <openssl/evp.h>
+#include <tss2/tss2_esys.h>
+
+#include "tpm2-tss-engine.h"
+#include "tpm2-tss-engine-common.h"
+
+#define HANDLE_SIZE 8
+
+static TPM2B_DATA allOutsideInfo = {
+    .size = 0,
+};
+
+static TPML_PCR_SELECTION allCreationPCR = {
+    .count = 0,
+};
+
+static TPM2B_PUBLIC keyTemplate = {
+    .publicArea = {
+        .type = TPM2_ALG_SYMCIPHER,
+        .nameAlg = ENGINE_HASH_ALG,
+        .objectAttributes = (TPMA_OBJECT_USERWITHAUTH |
+                             TPMA_OBJECT_SIGN_ENCRYPT |
+                             TPMA_OBJECT_DECRYPT |
+                             TPMA_OBJECT_FIXEDTPM |
+                             TPMA_OBJECT_FIXEDPARENT |
+                             TPMA_OBJECT_SENSITIVEDATAORIGIN |
+                             TPMA_OBJECT_NODA),
+        .authPolicy.size = 0,
+        .parameters.symDetail.sym = {
+             .algorithm = TPM2_ALG_NULL,
+             .keyBits.sym = 0,
+             .mode.sym = 0,
+              },
+        .unique.sym.size = 0
+     }
+};
+
+static int tpm2_cipher_nids[] = {
+    NID_aes_192_cbc,
+    NID_aes_256_cfb128,
+    NID_aes_256_cbc,
+    NID_aes_256_ofb128,
+    0
+};
+
+static int
+convert_array_hex_to_int(const unsigned char *in, size_t size)
```

Comment:
This naming is misleading, since it's not a hex array. Rather this is a `big-endian-array-to-native` conversion. My recommended code to also prevent page-align problems would be:
```
if (size != 4) fail;
uint32 r;
memcpy(&r, in, 4);
r = be32toh(r);
```

Reply:
Okay I needed to add the flag : AX_ADD_PREPROC_FLAG([-D_DEFAULT_SOURCE]) in order to use the function be32toh() with <endian.h>